asciidoc/volume1/use-cases/tf1-ch-c-use-case-stad.adoc

```diff
@@ -51,7 +51,29 @@ NOTE: The 50ms target accuracy is suitable for highly demanding use cases like r
 ====
 ****
 
+.R1521
+[sdpi_requirement#r1521,sdpi_req_level=should]
+****
+The <<term_manufacturer>> of a <<vol1_spec_sdpi_p_actor_somds_participant>> should configure its <<acronym_ts_service>> client to give priority to system clock adjustments that are slewing (over stepping adjustments).
+
+.Notes
+[%collapsible]
+====
+NOTE: If possible, the priority of slewing adjustments starts applying once the <<vol1_spec_sdpi_p_actor_somds_participant>> has acquired synchronization to the <<acronym_ts_service>> after a system (re)start.
+====
+****
 
+.R1522
+[sdpi_requirement#r1522,sdpi_req_level=shall]
+****
+When the <<vol1_spec_sdpi_p_actor_somds_provider>> detects a stepping adjustment of its system clock, the <<vol1_spec_sdpi_p_actor_somds_provider>> shall initiate a new MDIB sequence by assigning a new MDIB sequence identifier.
```

Comment:
We see the creation of a new MDIB sequence identifier critical - especially in conjunction with the SDC history service: how can the SDC consumer detect if the new sequence was caused due to a device reboot (data gaps), or a clock adjustment (no data gap)? Does the ClockState still provides the information, when the clock has changed? Is the information in the old or new sequence?

Reply:
@d-gregorczyk

Reply:
**SDPi Friday (3.11) Discussion:**  

1. Issue will never go away ... never be able to achieve 100% certainty that there will be no time slewing time adjustments
2. Alternatives?  Still under discussion ...
3. **RESOLUTION**:  (a) Keep R1522 as stated; (b) add note to requirement with Peter's concerns as a "warning"; (c) add a tech advisory requirement indicating the change (to address the last "gap detection" complexity question), including a specification of when the advisory should be sent with respect to the sequence "gap"; (d) [possibly] Add a 2nd advisory alert condition to indicate the _reason_ for a sequence ID change and "gap" in the data

Reply:
@PeterKranich, what do you think of the resolution ideas of today's SDPi call? If anything is unclear there feel free to reach out to me directly. Oh...and BTW, if (c) and (d) are implemented there may be limited to no need of (b).

Reply:
**SDPi Friday (10 Nov) Discussion:**

- Solutions above do not address the whole issue
- Alternative idea (direction) is to drop the "new sequence ID" requirement and introduce an extension to the clock state for the provider. Possibly keep an alternative with the "new sequence ID" requirement for those devices that cannot support the extension.
- Concluded that further discussion is needed, with @d-gregorczyk present. Probably at next week's call about component metrics, time allowing.

Reply:
On a meeting last Wednesday we came to the conclusion that non-slewing time adjustments should not cause a provider to trigger a sequence id change. Primarily, because time does not seem to be a key factor in clinical relevance, opposed to continuity of measurements. We agreed that rather firing a "clock changed" event can help in getting the situation under control.

I later re-thought this idea and was not sure anymore if an event is really doing the trick. My two cents:

1. A non-slewing time adjustment is an intrusive action that should not come up under normal conditions - normal conditions is slewing adjustment. Encountering non-slewing time adjustments indicate a malfunctioning NTP server or device clock - at least when the adjustment ensues far after a boot of the device. Doing a "soft reboot" after a subsequent non-slewing adjustment (initial ones after device start are ok) may be a reasonable action.
2. If sticking to the event approach, after an event is fired, what is the expected behavior regarding existing or new timestamps in the MDIB? As time can jump in any direction arbitrarily, not a single timestamp (except for those of continuous measurements perhaps) can be considered as being correct then. This accounts for any intermittent measurements, context binding times and alarm condition change timestamps. Bearing that in mind, when a non-slewing time adjustment occurs, and with that the transmission of an event, consumers can never trust in any timestamps until the sequence id changes anyway. Or there need to be some sophisticated, nifty re-calculation of timestamps for which I do not have any clue on where to start or how to proceed.

Reply:
I also talked with my colleges and they agree that creating a new sequence id would have additional implications in addition to the gaps in the contineous data (vitals and waves):

- Eperiodic data such as an NIBP: what should happen after the new sequence id change? Will the NIBP be invalidated? will the timestamp be corrected the new timestamp? -> this is a really bad idea if the NIBP has already been exported to other systems (EMR) and will then be exported once again with another timestamp.
- Alert events: what should happen with active alerts? Updating the timestamp and keeping the current occurence id? Or creating a new occurence id? -> this may cause trouble with the alert distribution systems

What my colleges recommend is to treat time like a parameter with a version #. The DeterminationTime would then marked with a time version number. When there is a non-slewing time adjustment, the version would increase and the consumer will get the information of the new time and the difference to the old time.
They pointed out that how these time changes are managed is very application specific and the SDC Provider should just provide all the information to the SDC cousumer which has to decide how to handle the situation.

Reply:
> Eperiodic data such as an NIBP: what should happen after the new sequence id change? Will the NIBP be invalidated? will the timestamp be corrected the new timestamp? -> this is a really bad idea if the NIBP has already been exported to other systems (EMR) and will then be exported once again with another timestamp.

The sequence id does not change under normal operating conditions. A non-slewing adjustment is an non-recoverable error - so we should treat as one.
 
> Alert events: what should happen with active alerts? Updating the timestamp and keeping the current occurence id? Or creating a new occurence id? -> this may cause trouble with the alert distribution systems

This is actually something that has to be considered in addition by SDPi or 11073-10702 respectively. As far as I can see we do not need to take it into consideration if we were doing a sequence id change, since a sequence id change implicates a breaking change to the system. Moreover, the APKP working group specified calculation of valid alarm limits imposed by particular standards based on the determination time. The calculation is only reasonable when time can be considered correct / strictly increasing.

> What my colleges recommend is to treat time like a parameter with a version #. The DeterminationTime would then marked with a time version number. When there is a non-slewing time adjustment, the version would increase and the consumer will get the information of the new time and the difference to the old time.

A time version number does not add any value in our case, it only makes things more complicated. First because consumers have to react upon sequence id changes anyway, and second because we can also use a latched advisory signal as proposed earlier, which signifies a consumer to not rely on any timestamps found in the MDIB.

## Bottom line

I think the only reasonable action is to perform a sequence id change. At least it is one valid option. Another one could be to mark all timestamps as tainted by using a latched alert signal. The use and modelling of that alert signal needs to be specified.

Reply:
Peter and I talked about potential solutions. We did not agree on a particular approach yet, but rather sketched the solution space:

1. Introduce an extension that allows for a SOMDS provider to signify timestamp validity - if the extension does not exist, the timestamps of a specific entity in the MDIB (aka metrics, alerts, contexts or other extensions that contain timestamps) can be considered up-to-date. We need to discuss if the timestamp validity extension expands to all children of device components or if every state receives its own extension (perhaps preferred as it does not break locality of reference**)
2. Use a latching alert signal - the signal would apply to the whole MDS or VMDs. No distinction between individual invalid timestamps of metrics and alerts.
3. Last resort: sequence id change which invalidates everything. This sledgehammer method is currently not preferred, but absolutely valid as sequence id changes can occur at any time - and consumers should be prepared to observe those sequence id changes anyway.

** In Software engineering, all relevant information for understanding a context should be localized at one point in the system and be comprehensible without knowledge of the context

Reply:
We had discussions at PAT#16 about the options to handle non-slewing time adjustments. We came to the conclusion that an extension may help in organizing timestamps. Each time segment between two non-slewing time adjusments should be versioned (see option 1). Timestamps of one segment can be compared to each other, timestamps of different segments can't be compared.

If a consumer does not want to rely on possibly outdated timestamps, it shall be able identify the latest segment and only use timestamps from that segment. If a consumer is not able to handle the extension at all, an alert signal may help in identifying if the time has been adjusted (see option 2) by a non-slewing event or not, and if there are still timestamps in the MDIB that may be invalid due to this adjustment.

*Enhanced solution of option 1:* Add deltas to time segments, so that consumers are able to compute the "latest time" based on the sum of deltas to the latest time. Caution: this solution should be thoroughly verified against feasibility and precision.

Reply:
> 1. Introduce an extension that allows for a SOMDS provider to signify timestamp validity - if the extension does not exist, the timestamps of a specific entity in the MDIB (aka metrics, alerts, contexts or other extensions that contain timestamps) can be considered up-to-date. We need to discuss if the timestamp validity extension expands to all children of device components or if every state receives its own extension (perhaps preferred as it does not break locality of reference**)
> 
> ** In Software engineering, all relevant information for understanding a context should be localized at one point in the system and be comprehensible without knowledge of the context

Is the context here the mdib (that is the mdib has a versioned clock with an e.g. ClockState/@TimeVersion) or every state update (that is, every AbstractState that has a time stamp gets an e.g., @TimeVersion)? It sounds like the later, but the former seems sufficient and in keeping with the rest of mdib updates. 

I noticed `pm:ClockState`  has both `@DateAndTime` and `@LastSet`. Could `@LastSet` be useful here if was to track only non-slewing time-adjustments (which includes the first time the clock is set)? Perhaps `@OperatingCycles` has value to indicate the number of slewing time adjustments that have been taken by the (clock) component as confirmation that clock remains synchronized at the provider end? Updating this count at "each" slewing adjustment would also provide (through `@DateAndTime`) the slew adjusted time. 

Finally, is a slewing period requirement going part of SDPi? I haven't been able to find a definitive source whether it is part of the NTP standard, which seems to have many non-slewing adjustments, but some sources indicate it could be system dependent and might be [300s](https://www.eecis.udel.edu/~mills/ntp/html/clock.html#step) or [600s](https://stackoverflow.com/questions/33746324/ntp-questions-slewing-versus-jumping-the-time). 5 - 10 minutes seems like a big clock error for SDC, but I'm not sure I have found the correct information.